h/services/user_delete.py

```diff
@@ -1,55 +1,247 @@
-import sqlalchemy as sa
+import logging
 
-from h.models import Annotation, Group, User
-from h.services.annotation_delete import AnnotationDeleteService
+from sqlalchemy import and_, delete, func, select, update
+
+from h.models import (
+    Annotation,
+    AuthTicket,
+    FeatureCohortUser,
+    Flag,
+    Group,
+    GroupMembership,
+    Job,
+    Token,
+    User,
+)
+
+log = logging.getLogger(__name__)
 
 
 class UserDeleteService:
-    def __init__(
-        self,
-        db_session: sa.orm.Session,
-        annotation_delete_service: AnnotationDeleteService,
-    ):
-        self._db = db_session
-        self._annotation_delete_service = annotation_delete_service
-
-    def delete_user(self, user: User):
+    def __init__(self, db, job_queue_svc, user_svc, limit=10000):
+        self.db = db
+        self.job_queue_svc = job_queue_svc
+        self.user_svc = user_svc
+        self.limit = limit
+        self.num_rows_deleted = 0
```

Comment:
It's convenient to store `num_rows_deleted` on `self` but it doesn't seem to work: it looks as if only one `UserDeleteService` object is created and that same one instance is re-used each time the periodic task runs.

In production this would presumably be one object per Celery worker. I think we only have one worker per instance (currently), but still it'd be a separate `UserDeleteService` object per instance.

And of course whenever an instance is recycled you'll get a new job (deployments, autoscaling, Gunicorn worker recycling).

This is certainly counter-intuitive so maybe we should fix it to create a new service object for each run of a Celery task.

Alternatively, each time `UserDeleteService.delete_user()` is called it could instantiate an instance of a helper class. That might be a good idea for code design anyway.

Reply:
I would say that, as a rule of thumb, services should be stateless, precisely to avoid this kind of side effects.

On the other hand, I've seen dependency injection frameworks where services could be scoped differently.

* Singleton: Once created, the same instance is kept in memory and re-used by following executions of the same worker.
* Request/job scope: Every time the worker processes a request/job, it creates new instances that are kept in memory until the end of the request/job. Following requests/jobs processed by the same worker will create new instances.
* Transient: A new instance is created every time, which causes less side effects and works more intuitively, but uses way more resources.

Would maybe Pyramid support something similar?

However, the problem is that this becomes very unintuitive as soon as singleton services start depending on other scoped services, so I would say we aim for stateless singletons as much as possible.